### ISSUE TITLE
prestop hook execution should respect grace period override

### ISSUE THREAD

**Issue by tedyu** [kubernetes/kubernetes#92432]
**What happened**:
Currently in kubeGenericRuntimeManager#killContainer, the grace period override is applied after the execution of prestop hook:
```
	if containerSpec.Lifecycle != nil && containerSpec.Lifecycle.PreStop != nil && gracePeriod > 0 {
		gracePeriod = gracePeriod - m.executePreStopHook(pod, containerID, containerSpec, gracePeriod)
	}
...
	if gracePeriodOverride != nil {
		gracePeriod = *gracePeriodOverride
		klog.V(3).Infof("Killing container %q, but using %d second grace period override", containerID, gracePeriod)
	}
```
If the gracePeriodOverride specifies smaller value (compared to pod.DeletionGracePeriodSeconds, e.g.), this means the override is ineffective
because the total duration from request to terminate until the actual shutdown would be longer than what gracePeriodOverride specifies.

**What you expected to happen**:
prestop hook execution should respect grace period override

**How to reproduce it (as minimally and precisely as possible)**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by tedyu** [comment]
[pod-del-grace-period-override.txt](https://github.com/kubernetes/kubernetes/files/4819897/pod-del-grace-period-override.txt)
Proposed fix

**Comment by rata** [comment]
@tedyu it is true that the `gracePeriodOverride` pointer is ignored for that. However, I'm not sure when that is used.

If you run `kubectl delete --grace-period 5 pod prestop` for example, it doesn't set `gracePeriodOverride` but it sets `pod.DeletionGracePeriodSeconds`.

To verify, I just added some log lines, here is the diff:
```diff
--- a/pkg/kubelet/kuberuntime/kuberuntime_container.go
+++ b/pkg/kubelet/kuberuntime/kuberuntime_container.go
@@ -571,16 +571,26 @@ func (m *kubeGenericRuntimeManager) killContainer(pod *v1.Pod, containerID kubec
                }
                pod, containerSpec = restoredPod, restoredContainer
        }
+       if gracePeriodOverride != nil {
+               klog.Errorf("gracePeriodOverride is %v", *gracePeriodOverride)
+       } else {
+               klog.Errorf("gracePeriodOverride is nil pointer")
+       }
 
        // From this point, pod and container must be non-nil.
        gracePeriod := int64(minimumGracePeriodInSeconds)
        switch {
        case pod.DeletionGracePeriodSeconds != nil:
                gracePeriod = *pod.DeletionGracePeriodSeconds
+               klog.Errorf("pod.DeletionGracePeriodSeconds is %v", pod.DeletionGracePeriodSeconds)
        case pod.Spec.TerminationGracePeriodSeconds != nil:
                gracePeriod = *pod.Spec.TerminationGracePeriodSeconds
        }
 
+       if pod.Spec.TerminationGracePeriodSeconds != nil {
+               klog.Errorf("pod.Spec.TerminationGracePeriodSeconds is %v", pod.DeletionGracePeriodSeconds)
+       }
+
        if len(message) == 0 {
                message = fmt.Sprintf("Stopping container %s", containerSpec.Name)
        }
```

And just running with a simple pod (I put it at the end), I can see that when I run `kubectl delete --grace-period 5 pod prestop` I see:
```
E0623 12:48:11.164673 1766314 kuberuntime_container.go:570] gracePeriodOverride is nil pointer
E0623 12:48:11.164680 1766314 kuberuntime_container.go:578] pod.DeletionGracePeriodSeconds is 5
E0623 12:48:11.164685 1766314 kuberuntime_container.go:584] pod.Spec.TerminationGracePeriodSeconds is 60
```

So, when using kubectl delete the grace period set on the command line is used for preStop hooks too. 

I honestly don't know if `gracePeriodOverride` is a leftover or if it's used in some place. Having a quick look, I didn't find a user. Does anyone already know if that param is unused?

But IIUC, changing the `killContainer()` function to respect the period wouldn't be backwards incompatible if there are no users (and hopefully they aren't, kubectl at least is not doing it). But if there are users, changing the time they used to had to execute might be more controversial.

Also, the code behaves as if `pod.DeletionGracePeriodSeconds` can not be set. I guess maybe the pod killer go routing might not set it, but haven't checked in detail.

I agree that if we have the `gracePeriodOverride` param, it will be nice to respect for preStop hooks. And having it will be useful for sidecar KEP, too, although there might be other alternatives for sidecars (like setting pod.DeletionGracePeriodSeconds **if that is possible**?).


What do you think?

**Comment by tedyu** [comment]
In the current sidecar implementation, the sidecar containers are killed following the termination of normal containers.
This may not give sidecar containers enough time to gracefully terminate.

**Comment by rata** [comment]
@tedyu not sure I follow, what do you mean with not having enough time to gracefully terminate? Even if it takes more than terminationGracePeriod seconds to kill a pod with sidecar containers, nothing will be interrupted, right? Or am I missing something?

**Comment by tedyu** [comment]
The termination of normal containers may exhaust the whole grace period, leading to direct killing of sidecar containers.

**Comment by tedyu** [comment]
@rata 
Does the above answer your question ?

**Comment by rata** [comment]
@tedyu oh, I understand now, thanks! My understanding, however, is that it doesn't do that (reading the KEP, not the implementation). If all the time is elapsed, 2 seconds (the `minimumGracePeriodInSeconds` as defined in `kuberuntime/kuberuntime_manager.go`) should be used.

**Comment by tedyu** [comment]
My comment yesterday was based on the following func
```
+func gracePeriodRemaining(pod *v1.Pod, gracePeriodOverride *int64, startTime time.Time) *int64 {
```
in #80744 which is called in terminating sidecars.

**Comment by Joseph-Irving** [comment]
For some context with regards to the Sidecar KEP, the intended behaviour is for Sidecars and Containers to share the Grace period, with the caveat that if the Containers use up all the grace period the sidecars will still be given the minimum (2 seconds) so that they don't get hard killed.

I used Grace period override as a way of recording the time elapsed by the Container termination so that I could pass the remaining time to the next Kill Container call for the sidecars. I did this as it just seemed like the easiest way to pass the time to each stage. 
I did not notice that the pre-stop hook call in killContainer does not respect the grace period override, I assumed that the override would be used throughout. So the effect of having a grace period that can extend to 3x the original was very much an unintended mistake.
I'm not entirely sure what the grace period override's intended use is, I just used it as it seemed convenient at the time, it does seem odd that it is ignored when it comes to the pre-stop hooks so it would be good if somebody with more knowledge about it's use could enlighten us.
